multigeiger/tube.cpp

```diff
@@ -35,22 +35,121 @@ TUBETYPE tubes[] = {
 
 // MUX (mutexes used for mutual exclusive access to isr variables)
 portMUX_TYPE mux_cap_full = portMUX_INITIALIZER_UNLOCKED;
 portMUX_TYPE mux_GMC_count = portMUX_INITIALIZER_UNLOCKED;
+portMUX_TYPE mux_hv = portMUX_INITIALIZER_UNLOCKED;
+
+void IRAM_ATTR reschedule(hw_timer_t *timer, unsigned int cycles) {
+  timerAlarmWrite(timer, cycles, false);  // set alarm after <cycles>, do not repeat
+  timerWrite(timer, 0);
+  timerAlarmEnable(timer);
+}
+
+// How many HV capacitor charge pulses to generate for any later charge cycle.
+#define MAX_CHARGE_PULSES 333
+
+// How many HV capacitor charge pulses to generate for first charge.
+// After device power on, HV capacitor is completely empty.
+#define MAX_CHARGE_PULSES_INITIAL 3333
+
+void IRAM_ATTR isr_recharge() {
+  static int state = 0;
+  static int charge_pulses;
+  static int max_charge_pulses = MAX_CHARGE_PULSES_INITIAL;  // only used once, see state 4
+  static unsigned int next;
+  if (state == 0) {
+    charge_pulses = 0;
+    portENTER_CRITICAL_ISR(&mux_cap_full);
+    isr_GMC_cap_full = 0;
+    portEXIT_CRITICAL_ISR(&mux_cap_full);
+    state = 1;
+    // fall through
+  }
+  while (state < 4) {
+    if (state == 1) {
+      digitalWrite(PIN_HV_FET_OUTPUT, HIGH);  // turn the HV FET on
+      state = 2;
+      reschedule(recharge_timer, 1500);  // 1500us (5000us gives 1.3 times more charge, 500us gives 1/20th of charge)
+      return;
+    }
+    if (state == 2) {
+      digitalWrite(PIN_HV_FET_OUTPUT, LOW);   // turn the HV FET off
+      state = 3;
+      reschedule(recharge_timer, 1000);  // 1000us
+      return;
+    }
+    if (state == 3) {
+      charge_pulses++;
+      if (isr_GMC_cap_full)
+        state = 4;
+      else if (charge_pulses < max_charge_pulses)
+        state = 1;
+      else
+        state = 5;
+      // fall through
+    }
+  }
+  if (state == 4) {
+    // capacitor full
+    portENTER_CRITICAL(&mux_hv);
+    isr_hv_charge_error = false;
+    isr_hv_pulses += charge_pulses;
+    portEXIT_CRITICAL(&mux_hv);
+    state = 0;
+    if (max_charge_pulses == MAX_CHARGE_PULSES_INITIAL) {
+      max_charge_pulses = MAX_CHARGE_PULSES;
+      next = 1000000;  // start with 1s interval
+    } else {
+      // depending on a lot of circumstances (e.g. level of radiation, humidity,
+      // leak currents, tube type, ...), we might need to charge the HV capacitor
+      // more or less often. we target charging with 2 pulses here.
+      if (charge_pulses <= 1) {
+        // one charge pulse was enough, so likely we charge too often
+        next = next * 5 / 4;
+      } else {
+        // 2 charge pulses: no change
+        // more: we charge too rarely, adapt in a linear way
+        next = next * 2 / charge_pulses;
+      }
+    }
+    // next = min(max(next, 1000), 10 * HVPULSE_MS * 1000)  // bounds: 1ms .. 10*HVPULSE_MS ms
+    reschedule(recharge_timer, next);
+    return;
+  }
+  if (state == 5) {
+    // capacitor does not charge! give up -> do not reschedule.
```

Comment:
Is this handled somewhere already with e.g. mandatory display on, warning sound etc.?
Otherwise we might leave the user baffled why it isn't working anymore. So maybe some reschedule in the far future (1 min? 10 min?) might be still helpful?

Reply:
there is only hv_error variable yet in main loop.